### DISCUSSION TITLE
Can I set per-layer underglow colors?

### DISCUSSION THREAD

**Discussion by Tohzt** [nickcoutsos/keymap-editor#251] [General]
For starters, this software is absolute tops. I could have learned to config zmk files manually but this was a great tool to stumble upon. As the title suggest, Im hoping to get some insight on how to set up custom underglow colors for each layer. I'm also having trouble learning how to set up individual key leds, so any pointers in that direction would be great

**Comment by Tohzt** [comment]
I have managed to get close. Using macros I can set a color when switching to a layout. But I would like to extend this to coloring my layout when using an &mo and have the color reset when releasing the &mo key

**Reply by Tohzt** [reply]
nevermind, that doesn't seem to actually set the color

**Reply by nickcoutsos** [reply]
Yeah, there's not a perfect solution I'm afraid. Internally (in ZMK, I mean) the layering and underglow systems are independent so the best you can do is have a macro that sets a colour and activates a layer, and uses `&macro_pause_for_release` to deactivate the layer and colour. A few problems here:

1. This would only trigger via this macro, so it wouldn't work with conditional layers and you'd need to use this macro in place of any binding to activate the layer.
2. Underglow colours aren't tracked so you can't just turn one off and "fall back" to the previous colour. You could either have the "off" state turn off underglow or switch to whatever animation/colour  you consider to be standard for your keyboard.

I remember someone working on a behavior to effectively trigger events for layers activating or deactivating, that would go some way towards solving the first problem so maybe that's an eventual solution if you're not bothered too much by the second part.

**Reply by Tohzt** [reply]
thanks for the info, i will look into getting some basic functionality from your suggestions. 

any thoughts in the per-key rgb's?
I'm not sure why, but turning on my underglow makes 4 keys on each half light up and those are the only per-key rgbs that light up. I don't see a way to customize color or effect

**Reply by nickcoutsos** [reply]
Sorry, I missed the part about per-key lighting. ZMK doesn't support that currently. There are animations but I don't think you can bind a key to activate a specific one, just cycle through enabled effects with `&rgb_ug RGB_EFF` (next effect) and `&rgb_ug RGB_RFR` (prev).
